### PR TITLE
Add crossover rendering

### LINKED ISSUE #513: Redraw tiles to make it obvious that you cannot change track at a crossover
These Vassal tiles (for a different game) should give you an idea of what I mean:

image

It's visually obvious that you can't change track at the crossover as one track lies on top of the other.  

Probably not necessary for experienced players, but as a noob I've been caught out by misreading the tiles. 

EDIT: The Carthaginian redesign came up with an elegant way of doing the crossovers: 

image

### PR DESCRIPTION
**Pull request by roseundy**
Fixes #513 

Also fixed boolean calls in path.rb and one random tile.

New rendering:
crossover
crossover_2
56_crossover
crossover_route

### CODE REVIEW

File: assets/app/view/game/part/track_node_path.rb [hunk 1]
```diff
@@ -256,6 +260,13 @@ def load_from_tile
             @end_x = edge_x_pos(@end_edge, 87)
             @end_y = edge_y_pos(@end_edge, 87)
             lanes = @path.lanes
+
+            if @tile.crossover? && @path.straight?
+              @crossover_dash = '1 55 63 56'
```
Comment (tobymao) [inline assets/app/view/game/part/track_node_path.rb:265]: these strings should be constants

File: assets/app/view/game/part/track_node_path.rb [hunk 2]
```diff
@@ -411,7 +428,19 @@ def render_part
             'stroke-dasharray': @dash,
           ) if @terminal
 
-          h(:path, props)
+          children = []
```
Comment (tobymao) [inline assets/app/view/game/part/track_node_path.rb:431]: children = [h(:path, props)]

File: assets/app/view/game/part/track_node_path.rb [hunk 3]
```diff
@@ -411,7 +428,19 @@ def render_part
...
+          children.append(h(:path, props))
+          if @crossover_dash
+            intersect_props = props.dup
```
Comment (tobymao) [inline assets/app/view/game/part/track_node_path.rb:434]: does this dup actually do what you want? dup is usually just a shallow copy, so nested dictionaries would still be mutated?

File: lib/engine/part/path.rb [hunk 4]
```diff
@@ -98,21 +98,55 @@ def path?
...
       def node?
-        @_node ||= @nodes.any?
+        return @_node if defined?(@_node)
+
+        @_node = @nodes.any?
       end
 
       def terminal?
-        @_terminal ||= !!@terminal
+        return @_terminal if defined?(@_terminal)
```
Comment (tobymao) [inline lib/engine/part/path.rb:107]: there's no reason to cache this, it's not expensive at all just do

def terminal?
!!@terminal

File: lib/engine/tile.rb [hunk 5]
```diff
@@ -360,6 +360,43 @@ def compute_city_town_edges
...
+    def crossover?
+      return @_crossover if defined?(@_crossover)
+
+      @_crossover = compute_crossover
+    end
+
+    def compute_crossover
+      edge_paths = Hash.new { |h, k| h[k] = [] }
+      @paths.each do |p|
+        edge_paths[p.a.num].append(p) if p.a.edge?
```
Comment (tobymao) [inline lib/engine/tile.rb:373]: use << in favor of append

File: assets/app/view/game/part/track_node_path.rb (hunk 3 above)
Comment (roseundy) [reply assets/app/view/game/part/track_node_path.rb:434]: It seems to work, but I can change to a a deep copy.

File: lib/engine/part/path.rb (hunk 4 above)
Comment (roseundy) [reply lib/engine/part/path.rb:107]: OK.

File: lib/engine/tile.rb (hunk 5 above)
Comment (roseundy) [reply lib/engine/tile.rb:373]: OK.

File: assets/app/view/game/part/track_node_path.rb (hunk 2 above)
Comment (roseundy) [reply assets/app/view/game/part/track_node_path.rb:431]: OK

File: assets/app/view/game/part/track_node_path.rb (hunk 1 above)
Comment (roseundy) [reply assets/app/view/game/part/track_node_path.rb:265]: OK.

File: lib/engine/tile.rb [hunk 6]
```diff
@@ -360,6 +360,43 @@ def compute_city_town_edges
...
+        edge_paths[p.b.num].append(p) if p.b.edge?
+        next unless p.nodes.size == 1
```
Comment (tobymao) [inline lib/engine/tile.rb:375]: net unless p.nodes.one?

File: lib/engine/tile.rb [hunk 7]
```diff
@@ -360,6 +360,43 @@ def compute_city_town_edges
...
+
+        ct_edge = preferred_city_town_edges[p.nodes.first]
```
Comment (tobymao) [inline lib/engine/tile.rb:377]: what's the point of these 3 lines? they don't do anything

File: lib/engine/tile.rb [hunk 8]
```diff
@@ -360,6 +360,43 @@ def compute_city_town_edges
...
+        p.straight?(ct_edge)
+        p.gentle_curve?(ct_edge)
+      end
+
+      @paths.each do |p|
+        next if p.nodes.size > 2
```
Comment (tobymao) [inline lib/engine/tile.rb:383]: how can a path have more than two nodes?

File: lib/engine/tile.rb (hunk 6 above)
Comment (roseundy) [reply lib/engine/tile.rb:375]: OK.

File: lib/engine/tile.rb (hunk 7 above)
Comment (roseundy) [reply lib/engine/tile.rb:377]: It calls those methods in path with the proper city/town thus causing them to cache. I don't want to call preferred_city_town_edges in Path.

File: lib/engine/tile.rb (hunk 8 above)
Comment (roseundy) [reply lib/engine/tile.rb:383]: Damn. Should be 1.

File: assets/app/view/game/part/track_node_path.rb (hunk 3 above)
Comment (tobymao) [reply assets/app/view/game/part/track_node_path.rb:434]: maybe separate out the hashes

File: lib/engine/tile.rb (hunk 7 above)
Comment (tobymao) [reply lib/engine/tile.rb:377]: what's the point of caching them here? why don't you just do it lazily?

Comment (tobymao) [reply lib/engine/tile.rb:377]: ohhhhh, wait what, you rely on the order for caching? that's a really dangerous pattern. the cache should take into consideration arguments

File: assets/app/view/game/part/track_node_path.rb (hunk 3 above)
Comment (roseundy) [reply assets/app/view/game/part/track_node_path.rb:434]: Actually I don't need to do either. It seems each call of h(:path, props) is essentially a call by value of the properties, so I'm free to change props between calls.

File: lib/engine/tile.rb (hunk 7 above)
Comment (roseundy) [reply lib/engine/tile.rb:377]: Revamped this to not use arguments at all and to avoid this entire problem.

**Review by tobymao — APPROVED** [review]

**Pull request merged**